### ISSUE TITLE
walk ast helper and ast postprocessing option

### ISSUE THREAD

**Issue by RobertDober** [pragdave/earmark#398]
Today:
implementing a custom transformer means doing the following, in pseudo code
```elixir
   ast = EarmarkParser.as_ast(markdown)
   my_transformation(ast)
```
where `mytransformation`  has no access to built in helpers or transformers

Tomorrow:

`Earmark`  will expose visitors and reducer to offer an event based or purely functional traversal

E.g.

```elixir
   my_transformer = fn {"p", atts, content, meta} -> {p, Keyword.put(atts, "data-source", "my_transformer"), content, meta) 
                                   other -> other end
   Earmark.map_ast(ast, my_transformer)
                               
end
```
allowing not to have a catch all pattern in the mapper (slower)

```elixir
 
  my_transformer = fn {"p", atts, content, meta} -> {p, Keyword.put(atts, "data-source", "my_transformer"), content, meta) end 
   Earmark.map_ast!(ast, my_transformer)
``` 

For structural change examples, please stby.

**Comment by josevalim** [comment]
Hi @RobertDober, the approach here is to usually have a `traverse` function that works like a reduce.

    EarmarkParser.traverse_ast(ast, acc, fn node, acc -> ... end)

The nice thing about the traverse is that if you want to traverse the ast only to collect some information and not modify it, you can!  But if all you need it as a map you can have this:

    def traverse_ast(ast, mapper) do
      {res, []} = traverse_ast(ast, [], fn node, [] -> {mapper.(node), []} end)
      res
    end

Also, I wouldn't worry about `map_ast!`, rescuing user code will always be complicated to say if the failure happened when immediately executing the function or some further inside the function. I think it is fine to tell users they need to add a catch-all clause. This is the approach done with Elixir's AST too.

**Comment by RobertDober** [comment]
Thank you @josevalim seems the triviality of the implementation of the map eluded me :wink

**Comment by RobertDober** [comment]
@josevalim just had some time to review the code, I did some work to assure that `walk_ast` was not recursive, I'll need some time to come up with a non recursive reducer... should be doable ;)

**Comment by joisig** [comment]
Not sure it's useful but in case it is, I'm a happy user of Earmark 1.4.19 (will upgrade soon!) and implemented the following walk-and-modify function (public domain, feel free to use however you like):

```
  @doc """
  Walks an AST and allows you to process it (storing details in acc) and/or
  modify it as it is walked.

  The process_item_fn function is required. It takes two parameters, the
  single item to process (which will either be a string or a 4-tuple) and
  the accumulator, and returns a tuple {processed_item, updated_acc}.
  Returning the empty list for processed_item will remove the item processed
  the AST.

  The process_list_fn function is optional and defaults to no modification of
  items or accumulator. It takes two parameters, the list of items that
  are the sub-items of a given element in the AST (or the top-level list of
  items), and the accumulator, and returns a tuple
  {processed_items_list, updated_acc}.

  This function ends up returning {ast, acc}.
  """
  def walk_and_modify_ast(items, acc, process_item_fn, process_list_fn \\ &({&1, &2}))
  when is_list(items) and is_function(process_item_fn) and is_function(process_list_fn)
  do
    {items, acc} = process_list_fn.(items, acc)
    {ast, acc} = Enum.map_reduce(items, acc, fn (item, acc) ->
      {_item, _acc} = walk_and_modify_ast_item(item, acc, process_item_fn, process_list_fn)
    end)
    {List.flatten(ast), acc}
  end

  def walk_and_modify_ast_item(item, acc, process_item_fn, process_list_fn)
  when is_function(process_item_fn) and is_function(process_list_fn) do
    case process_item_fn.(item, acc) do
      {{type, attribs, items, annotations}, acc}
      when is_binary(type) and is_list(attribs) and is_list(items) and is_map(annotations) ->
        {items, acc} = walk_and_modify_ast(items, acc, process_item_fn, process_list_fn)
        {{type, attribs, List.flatten(items), annotations}, acc}
      {item_or_items, acc} when is_binary(item_or_items) or is_list(item_or_items) ->
        {item_or_items, acc}
    end
  end

```

**Comment by RobertDober** [comment]
@joisig as mentioned in the ML I will look at this ASAP but just a question, is this not already possible (`Earmark.Transformer.map_ast_with/3 and 4)[https://github.com/pragdave/earmark/blob/947b9acbc446c32433098f5118bb6921773213c2/lib/earmark/transform.ex#L309]

**Comment by joisig** [comment]
Hi @RobertDober I may have misunderstood, but I didn't think that function and the related ones would let you modify the structure of the AST, whereas what I suggested does (for an example see https://elixirforum.com/t/earmark-elixirs-markdown-converter/25208/112). Sorry if I wasted your time, and of course no rush to incorporate any of this, just wanted to provide it for anybody who might want to use it as-is.

**Comment by RobertDober** [comment]
No you are right there are no structure modifying transformations, but I was thinking about them, I will read the forum post in detail

sounds interesting

bye for now

**Comment by RobertDober** [comment]
@joisig this is definitely worth a Pull Request, ideally with some tests.

May I make some suggestions (but I can implement them myself if you do not have time)

- do not use the guard clauses `is_function` on recursive calls, e.g.

  ```elixir
        def public_fn(args) when some_condition(args), do: private_fn(args)

        def private_fn(args) do # contains the recursive calls, that do not need the guards anymore
   ```

- replace `{_i, _a} = some_call(...)` by `some_call(...)`
- add a catch it all error case to the `case process_item_fn.(item, acc) do` as we call a client function I would prefer
  to raise a more meaningfull error

Thank you for your work

**Comment by RobertDober** [comment]
ah about the last point, seems José does not agree with me ;)

**Comment by joisig** [comment]
Sure, will be happy to work this up into a pull request with tests and addressing your comments, and send to you in the next few days.

**Comment by joisig** [comment]
@RobertDober see https://github.com/pragdave/earmark/pull/454

**Comment by RobertDober** [comment]
I do not understand @joisig your PR has been merged already here #455 did I miss something?

**Comment by RobertDober** [comment]
> Hi @RobertDober I'm not sure - I thought you had created #455 and merged it? The one I created was #454 but GitHub tells me that in #455 my branch has in fact been merged into your branch pragdave:i398-restructuring-ast-traversal

indeed I will probably move your code into a different module and might refactor before merging it into master again, this will not devaluate your contribution which is also duely noted in the `RELEASE.md`
hope you are fine with this

**Comment by joisig** [comment]
@RobertDober I'm absolutely fine with that, thanks!
